packages/hash/frontend/src/components/hooks/blockProtocolFunctions/knowledge/useBlockProtocolArchiveEntity.ts

```diff
@@ -0,0 +1,58 @@
+import { useMutation } from "@apollo/client";
+import { useCallback } from "react";
+
+import {
+  ArchiveEntityMutation,
+  ArchiveEntityMutationVariables,
+} from "../../../../graphql/apiTypes.gen";
+import { archiveEntityMutation } from "../../../../graphql/queries/knowledge/entity.queries";
+import { ArchiveEntityMessageCallback } from "./knowledge-shim";
+
+export const useBlockProtocolArchiveEntity = (
+  readonly?: boolean,
+): {
+  archiveEntity: ArchiveEntityMessageCallback;
+} => {
+  const [archiveEntityFn] = useMutation<
+    ArchiveEntityMutation,
+    ArchiveEntityMutationVariables
+  >(archiveEntityMutation, {
+    /** @todo reconsider caching. This is done for testing/demo purposes. */
+    fetchPolicy: "no-cache",
+  });
+
+  const archiveEntity: ArchiveEntityMessageCallback = useCallback(
+    async ({ data }) => {
+      if (readonly) {
+        return {
+          errors: [
+            {
+              code: "FORBIDDEN",
+              message: "Operation can't be carried out in readonly mode",
+            },
+          ],
+        };
+      }
+
+      if (!data) {
+        return {
+          errors: [
+            {
+              code: "INVALID_INPUT",
+              message: "'data' must be provided for archiveEntity",
+            },
+          ],
+        };
+      }
+
+      const { entityId } = data;
+
+      await archiveEntityFn({ variables: { entityId } });
```

Comment:
Do you not need to use the return value here?

Reply:
No, because I've made the return-type `Promise<void>`

Reply:
We seem to throw errors when `data` is nullish in other hooks so I feel like we need to check it's true here?

Reply:
Why is `data` nullable in other hooks?

Reply:
I feel like the error throwing when `data` is falsy is a consequence of an incorrect return-type, so not something that should influence the decision of which return-type to use here.

Reply:
It doesn't seem to be our configuration, but actually hardcoded in the type from Apollo 

<img width="452" alt="image" src="https://user-images.githubusercontent.com/25749103/204308459-ad4d0fc5-226b-436a-b00d-64ff46dcfc73.png">

Reply:
I think `data` can be `undefined` in apollo, because they return errors rather than throwing them (see the `error` field in your screenshot). I now see that this is what we are doing for the BP hooks as well:
<img width="939" alt="image" src="https://user-images.githubusercontent.com/42802102/204341668-8cfef9c2-a5c8-4854-9a92-977ae9374759.png">

However, I still don't think returning `data: undefined` should be treated as an error state, instead if `errors` is a non-empty array it should indicate that an error has occurred in the `archiveEntity` hook.

Reply:
I've looked at the current graph service definition, and it seems we should be returning a boolean instead so have changed it to that in https://github.com/hashintel/hash/pull/1518/commits/6f3978a1110a0d7a4a8e3b21f090a70d1adc2fc6
<img width="474" alt="image" src="https://user-images.githubusercontent.com/42802102/204342465-a0febca9-5bd1-412d-a5d5-a63513f520e7.png">